### ISSUE TITLE
Issue using Docker volumes with read replicas

### ISSUE THREAD

**Issue by danielkermode** [docker-library/mysql#292]
I feel like I've seen this error before but couldn't find anything with a Google search, so if someone could point me in the right direction that would be great.

I start my Docker containers which are master/slave MySQL databases through Docker Swarm/Compose with `docker stack deploy`. In my `.yml` file, I have volumes set up on the master and slave with
```yml
volumes:
  - '/storage/docker/mysql:/var/lib/mysql'
```

I configure the databases to be master/slave manually with `mysql` shells in each Docker container.

Unfortunately, when I run `START SLAVE;` on the slave, the following error occurs:

`ERROR 1872 (HY000): Slave failed to initialize relay log info structure from the repository`

Not using volumes on the slave db service (but still using them on the master) fixes this error and the slave starts and works correctly.

Is this a Docker issue or am I missing something?

**Comment by ltangvald** [comment]
I don't know this issue well, though I've seen some related reports regarding the slaves having crashed at some point. You might look through https://bugs.mysql.com/bug.php?id=77496 (it's marked as closed, but the final comments mention seeing the issue in 5.7.16, and has a suggested workaround)

**Comment by danielkermode** [comment]
Thanks, I've decided to not use Docker with MySQL in production as there's too many workarounds for it to be worth the effort. Replication seems to be a use case that isn't fully supported yet (I looked at some other MySQL images that tried to solve this but they had their own issues). Docker's "stateless" philosophy really doesn't seem to gel with databases at this stage, although it's fine for development.

**Comment by bkuhl** [comment]
This issue should be closed.

**Comment by Nicnl** [comment]
Hello all,

Sorry for digging out such an old issue (three years now)  
But I've spent the last last two hours trying to setup master/slave replication on MySQL containers without any luck  
I had the same error each time the MySQL container was restarted  
**And unfortunately the one and only google result talking about it... is here**  
So, again, pardon me for digging out such an old thread;  
But I feel like I had to write about this somewhere, and this thread seems to be the most useful place  

**Tldr:**  
It's not a bug ~~, it's a feature~~; Fortunately the solution isn't a workaround either  
It's just an unfortunate conflict between MySQL's default settings and Docker's behaviour  
The solution is at the end if don't care about the cause

________

## Why is this happening

The error message is the following:   
`ERROR 1872 (HY000): Slave failed to initialize relay log info structure from the repository`  
It's convoluted, but it indicates that there's something wrong with the relay files that the slave uses for synchronizing with the master

Let's take a look at my `/var/lib/mysql` directory:   
```bash
bash-4.2$ ls -alh
total 271M
drwxr-xr-x 6 1000 1000 4.0K Jun 25 09:23 .
drwxr-xr-x 1 root root 4.0K May 21 04:36 ..
-rw-r----- 1 1000 root  214 Jun 25 09:14 1b41a0de71e8-relay-bin.000001
-rw-r----- 1 1000 root  49K Jun 25 09:14 1b41a0de71e8-relay-bin.000002
-rw-r----- 1 1000 root   64 Jun 25 09:14 1b41a0de71e8-relay-bin.index
-rw-r----- 1 1000 root  177 Jun 25 09:23 3a2f9d05ef62-relay-bin.000001
-rw-r----- 1 1000 root   32 Jun 25 09:23 3a2f9d05ef62-relay-bin.index
-rw-r----- 1 1000 root  27K Jun 25 09:02 9299f99bd991-relay-bin.000002
-rw-r----- 1 1000 root 9.0K Jun 25 09:06 9299f99bd991-relay-bin.000003
-rw-r----- 1 1000 root   64 Jun 25 09:02 9299f99bd991-relay-bin.index
...
```

The problem is right in front of our eyes;  
For some reason, there's multiple relay logs files

Here's the important part from the official documentation on slave relay logs files:  
&nbsp;&nbsp;&nbsp;_(https://dev.mysql.com/doc/refman/5.7/en/slave-logs-relaylog.html)_    
```
By default, relay log file names have the form host_name-relay-bin.nnnnnn in the data directory,
where host_name is the name of the slave server host and nnnnnn is a sequence number.
```

Remember that new docker containers have... randomized hostnames?  
So, yeah... each time my slave container was restarted, a new hostname was generated (12 random hex characters)   
And that means that the slave tried to use a non-existent relay log file.  

It's silly. I've spend almost three hours on it. But it's such a relief.

## How to fix it

On your slave, you have to configure a hostname that's not going to change.  

For instance, with docker-compose, you just have to add  
&nbsp;&nbsp;`hostname: 'somethingsomething'`  

## Message to the maintainers

Sorry for bothering you @yosifkit but I don't know who's in charge and who I ough to ping

I think it would be a good idea to add a few lines about this in your docker readme page,  
&nbsp;&nbsp;recommending to either set a definitive hostname  
(this or simply change the mysql configuration so that the relay log file does not contains the hostname)

Since I'm not using Percona and not the official mysql docker container, that would be a great addition to their README as well  
(I hope it'll work)  Pinging @delgod since you seem to be part of the Percona organization
